brushfire-core/src/main/scala/com/stripe/brushfire/Brushfire.scala

```diff
@@ -5,21 +5,30 @@ import com.twitter.algebird._
 /**
  * Represents a single instance of training data.
  *
- * @tparam K feature names
- * @tparam V feature values
+ * @tparam M instance metadata
+ * @tparam F features
  * @tparam T target distribution
  *
  * @constructor create a new instance
- * @param id an identifier unique to this instance
- * @param timestamp the time this instance was observed
+ * @param metadata metadata about this instance
  * @param features a map of named features that make up this instance
  * @param target a distribution of predictions or labels for this instance
  */
-case class Instance[K, V, T](id: String, timestamp: Long, features: Map[K, V], target: T)
+case class Instance[M, F, T](metadata: M, features: F, target: T) {
+  def id(implicit eq: M =:= DefaultMetadata): String = {
```

Comment:
I understand adding these for source-level compatibility, but I think I'd rather just make this a breaking change (we can bump the version number appropriately) and get rid of the special casing here - or rather, push it to the places that actually need it.